### ISSUE TITLE
Inquiry About Norm Ratio Calculation in Your Paper

### ISSUE THREAD

**Issue by hiroi-sora** [JunyiZhu-AI/surrogate_model_extension#2]
Your paper states: ( in 5.1 Low-Rank Property of Local Steps )

> To support `Assumption 4.2` empirically, we collect the true gradients of the local training steps $\lbrace\nabla w_t\rbrace_{t=0}^{T-1}$ and compute the two principal components with the largest two singular values. We let $P_2$ consist of these two principal components. Then we project the gradients $\lbrace\nabla w_t\rbrace_{t=0}^{T-1}$ onto $V_2$ using $P_2$ and compute the norm ratio $\lbrace \|\|P_2\nabla w_t\|\|/\|\|\nabla w_t\|\|\rbrace_{t=0}^{T-1}$.

I have attempted to replicate this part as per the description. Below is the testing code based on this repository:

```python
def train(net, trainloader, epochs, train_lr, setup):
    net.eval()
    criterion = torch.nn.CrossEntropyLoss(reduction="mean")
    optimizer = torch.optim.SGD(params=net.parameters(), lr=train_lr)
    gradients = []  # nabla_wt
    for _ in range(epochs):
        for data, label in trainloader:
            optimizer.zero_grad()
            data = data.to(**setup)
            label = label.to(device=setup["device"])
            pred = net(data)
            loss = criterion(input=pred, target=label)
            loss.backward()

            ### collect the true gradients of the local training steps ###
            grad = []
            for param in net.parameters():
                if param.grad is not None:
                    grad.append(param.grad.view(-1))
            grad = torch.cat(grad)
            gradients.append(grad.cpu().data.numpy())

            optimizer.step()

    ### compute the two principal components with the largest two singular values ###
    pca = PCA(n_components=2)  # sklearn.decomposition
    pca.fit(np.array(gradients))
    P2 = pca.components_  # let P2 consist of these two principal components
    ### compute the norm ratio ###
    print("||P2 ∇wt|| /  ||∇wt||")
    for grad in gradients:
        norm_p2_nabla_wt = np.linalg.norm(np.dot(P2, grad.T))  # || P2 ∇wt ||
        norm_nabla_wt = np.linalg.norm(grad)  # || ∇wt ||
        norm_ratio = norm_p2_nabla_wt / norm_nabla_wt
        print(norm_ratio)

```

However, I found that the calculated norm_ratio is less than 0.5, which seems inconsistent with the conclusions depicted in the figures of your paper.

![image](https://github.com/JunyiZhu-AI/surrogate_model_extension/assets/56373419/52ed0921-4331-4042-9fa5-d030b80cc58f)

Could there be an error in my testing code? I would greatly appreciate any guidance or corrections you might suggest.

**Comment by hiroi-sora** [comment]
Indeed, there was an error in my code. When calculating $P_2$, PCA should not be used. Instead, use SVD (and do not centered).

The following code can replicate the conclusions of the paper:

```python
    svd = TruncatedSVD(n_components=2)  # sklearn.decomposition
    svd.fit(np.array(gradients))
    P2 = svd.components_
```

Or use torch on GPU:

```python
    gradients_tensor = torch.stack(gradients)  # gradients = [tensor]
    _U, _S, V = torch.linalg.svd(gradients_tensor, full_matrices=False)
    P2 = V[:2, :]
```

**Comment by JunyiZhu-AI** [comment]
Good to know that this issue is addressed, and thanks for sharing your insight!
